Stride.BepuPhysics/Processors/ContainerProcessor.cs

```diff
@@ -91,29 +92,29 @@ public override void Update(GameTime time)
                     bepuSim.RemainingUpdateTime -= bepuSim.SimulationFixedStep;
                     stepCount++;
                 }
-                //simStepWatch.Stop();
-
-                //parForWatch.Start();
 
+                #warning I don't think this should be user-controllable ? We don't provide control over the other parts of the engine when they run through the dispatcher and having it on or of doesn't (or rather shouldn't) actually change the result, just how fast it resolves
+                // I guess it could make sense when running on a low power device, but at that point might as well make the change to Dispatcher itself
                 if (bepuSim.ParallelUpdate)
                 {
                     Dispatcher.For(0, bepuSim.Simulation.Bodies.ActiveSet.Count, (i) =>
                     {
                         var handle = bepuSim.Simulation.Bodies.ActiveSet.IndexToHandle[i];
-                        var BodyContainer = bepuSim.BodiesContainers[handle];
+                        var bodyContainer = bepuSim.BodiesContainers[handle];
                         var body = bepuSim.Simulation.Bodies[handle];
 
-                        var entityTransform = BodyContainer.Entity.Transform;
-                        entityTransform.WorldMatrix.Decompose(out Vector3 _, out Quaternion _, out Vector3 ContainerWorldTranslation);
-                        var ParentEntityTransform = new Vector3();
-                        var parent = BodyContainer.Entity.GetParent();
+                        var parentEntityTransform = new Vector3();
+                        var parent = bodyContainer.Entity.Transform.Parent;
                         if (parent != null)
                         {
-                            parent.Transform.WorldMatrix.Decompose(out Vector3 _, out Quaternion _, out ParentEntityTransform);
+                            parent.WorldMatrix.Decompose(out Vector3 _, out Quaternion _, out parentEntityTransform);
                         }
 
-                        entityTransform.Position = body.Pose.Position.ToStrideVector() - BodyContainer.CenterOfMass - ParentEntityTransform;
+                        var entityTransform = bodyContainer.Entity.Transform;
+                        #warning fix this part, conversion from local to world is just halfway done here
+                        entityTransform.Position = body.Pose.Position.ToStrideVector() - bodyContainer.CenterOfMass - parentEntityTransform;
                         entityTransform.Rotation = body.Pose.Orientation.ToStrideQuaternion();
+                        #warning this operation is not thread safe if multiple containers are in the same hierarchy, perhaps best to avoid calling this method and instead let the engine update world matrix itself, maybe by moving this processor right before that process
```

Comment:
would this be an issue if UseTRS was not enabled?

Reply:
Yes, the issue is that this function writes to the matrices of that component, but also to the matrices of its parents.
So, if a container has a parent which is also a container, two different threads may process them at the same time, reading and writing to the same matrices, from partially set/not fully ready matrix state.